src/type.ts

```diff
@@ -157,30 +190,31 @@ export interface TypeVisit<P, R> {
 
 export function type_visit<P, R>(visitor: TypeVisit<P, R>,
                           type: Type, param: P): R {
-  if (type instanceof PrimitiveType) {
-    return visitor.visit_primitive(type, param);
-  } else if (type instanceof FunType) {
-    return visitor.visit_fun(type, param);
-  } else if (type instanceof CodeType) {
-    return visitor.visit_code(type, param);
-  } else if (type instanceof AnyType) {
-    return visitor.visit_any(type, param);
-  } else if (type instanceof VoidType) {
-    return visitor.visit_void(type, param);
-  } else if (type instanceof ConstructorType) {
-    return visitor.visit_constructor(type, param);
-  } else if (type instanceof InstanceType) {
-    return visitor.visit_instance(type, param);
-  } else if (type instanceof QuantifiedType) {
-    return visitor.visit_quantified(type, param);
-  } else if (type instanceof VariableType) {
-    return visitor.visit_variable(type, param);
-  } else if (type instanceof OverloadedType) {
-    return visitor.visit_overloaded(type, param);
-  } else if (type instanceof TupleType) {
-    return visitor.visit_tuple(type, param);
-  } else {
-    throw "error: unknown type kind " + typeof(type);
+  switch (type.type) {
+    case TypeType.PRIMITIVE:
+      return visitor.visit_primitive(type, param);
+    case TypeType.FUN:
```

Comment:
I think you also need a `VARIADIC_FUN` case here too. (Previously, `instanceof FunType` worked for both types.)

Reply:
Good point! And now this is something we can catch by checking for exhaustiveness…